Add automember module

jake2184 opened the pull request:
I've extended & updated an automember module, first proposed in #392. That PR looked active for a couple of weeks, but silent for months. I've forked it, rebased on master, and addressed a few of the issues highlighted in the PR.

I can't tag the previous author, Mark Hahl / @Wolskie. I haven't changed any of the author notes or copyright.

jake2184 commented:
oh, the tag did work! Just didn't autocomplete.

rjeffman commented:
Hello @jake2184,

The task failing is due to differences in string handling in Python 2 and 3. To fix this issue you should add the following to the imports:

```python
from ansible.module_utils._text import to_text
```

And then, when you need the string, you use `to_text(data)`

Thank you for you contribution!

jake2184 commented:
I seem to be getting errors now that I'm fairly sure are unrelated to any of my changes: 
```
> {"ansible_facts": {"discovered_interpreter_python": "/usr/libexec/platform-python"}, "changed": false, "msg": "Kerberos authentication failed: kinit: Cannot read password while getting initial credentials\n"
```

This is during facts collection? Is there anything I need to setup (rebase on master?) or change to get these working?

rjeffman commented:
That is really weird... I'll look into that in the next days, as I really don't have the time now.

rjeffman commented on playbooks/automember/automember-absent.yml:
```diff
@@ -0,0 +1,18 @@
+      type: group
+      state: absent
+
+  - name: Ensure hostgroup automember rule ipaservers is absent
```
Please, separate this playbook in two, one for `group`, one for `hostgroup`.

jake2184 replied:
Done

rjeffman commented on playbooks/automember/automember-present.yml:
```diff
@@ -0,0 +1,18 @@
+      type: group
+      state: present
+
+  - name: Ensure hostgroup automember rule ipaservers is absent
```
Please, separate this playbook in two, one for group, one for hostgroup.

jake2184 replied:
Done

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,388 @@
+
+
+from ansible.module_utils._text import to_text
+from ansible.module_utils.ansible_freeipa_module import (api_command,
```
Either format like `black`, or (better yet) add the list of imports in the least number of lines possible.

If you start the list in the second line, with a `tab`, it will work:
```python
from package import (
    method01, method02, method03, method04,
    methdo05
)
```

jake2184 replied:
Done

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,388 @@
+    state = ansible_module.params.get("state")
+
+    # grouping/type
+    grouping = ansible_module.params.get("type")
```
Since `type` doesn't work well for variable name, in ansible-freeipa we use `<what>_type`, in this case `grouping_type` or even `automember_type`. Also use this name as attribute in the YAML file. Please, do not use `type`.

jake2184 replied:
Done

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,388 @@
+                                 {"type": to_text(grouping)}])
+
+        errors = []
+        for name, command, args in commands:
```
ansible-freeipa modules must support `check_mode`. Check file [utils/templates/ipamodule+member.py.in](https://github.com/rafasgj/ansible-freeipa/blob/masterutils/templates/ipamodule+member.py.in), [line 289](https://github.com/freeipa/ansible-freeipa/blob/0cfd07a70966e5e43975c415be0445a774a1cf45/utils/templates/ipamodule%2Bmember.py.in#L289) to see how this is being implemented.

jake2184 replied:
Done

rjeffman commented on tests/automember/test_automember.yml:
```diff
@@ -0,0 +1,162 @@
+
+  # CLEANUP TEST ITEMS
+
+  - name: Ensure group automember rule admins is absent
+    ipaautomember:
+      ipaadmin_password: SomeADMINpassword
+      name: admins
+      state: absent
+      type: group
+
+  - name: Ensure hostgroup automember rule ipaservers is absent
+    ipaautomember:
+      ipaadmin_password: SomeADMINpassword
+      name: ipaservers
+      state: absent
+      type: hostgroup
```
Please, and a testing group and hostgroup (with `ipagroup` and `ipahostgroup`), so that the whole group or hostgroup can be cleaned up before and after test is executed. (`testgroup` and `testhostgroup` are great names, in this case.)

jake2184 replied:
Done

rjeffman commented:
@jake2184 the failures on the checks, currently, have nothing to do with your changes as we are having some issues with the images used for testing. I'm fixing those, but we can't use the checks for a while. I'll let you know when everything is working again.

rjeffman commented:
@chr15p, all test environments are up and running again. Please, fix the failing tests.

rjeffman commented:
@chr15p, all test environments are up and running again. Please, fix the failing tests.

jake2184 commented:
Tests running cleanly

rjeffman commented:
@jake2184 great job!

I'll try to do a proper review during this week.

jake2184 commented:
Can I bump this please @rjeffman ?

jake2184 commented:
And again @rjeffman @t-woerner

rjeffman commented:
@jake2184 I was on vacation, and only returning "tomorrow".

I'll look at it Really Soon Now (tm). :-)

jake2184 commented:
No worries, as long as it doesn't fall through the cracks :)

jake2184 commented:
Is this going to be looked at any time soon? @rjeffman

rjeffman commented:
Hello, sorry for the too long delay.

We're having issues with CI (and available time), as soon as CI is fixed, we'll restart pushing PRs, automount and automember modules are a priority.

I hope this to be soon, but can't promise as, right now, **time** is at a premium.

rjeffman commented:
Hello @jake2184,

We had issues with CI, and the fixes require that all pending PR are rebased.

Can you please rebase your PR?

jake2184 commented:
Merged master in, pending CI checks

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,385 @@
+    state = ansible_module.params.get("state")
+
+    # grouping/type
+    grouping = ansible_module.params.get("automember_type")
```
Please, change `grouping` to `automember_type`. Try to use variable names as close as possible to the attribute names.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,385 @@
+    return transformed
+
+
+def gen_condition_commands(name,
+                           grouping,
+                           module_conditions,
+                           current_conditions):
+    """Return a list of commands to add/remove automember rule conditions."""
+    commands = []
+    add_diff = set(module_conditions) - set(current_conditions)
+    remove_diff = set(current_conditions) - set(module_conditions)
+
+    for item in add_diff:
+        key, condition = item.split("=")
+        condition_args = gen_condition_args(grouping, key, condition, None)
+        commands.append([name, 'automember_add_condition', condition_args])
+
+    for item in remove_diff:
+        key, condition = item.split("=")
+        condition_args = gen_condition_args(grouping, key, condition, None)
+        commands.append([name, 'automember_remove_condition', condition_args])
+
+    return commands
```
This function is not used anywhere.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,385 @@
+                    exclusive_add = transform_conditions(exclusive or [])
+                    exclusive_del = []
+
+                if inclusive is not None:
```
This test is not needed as you will be using `inclusive_*`.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,385 @@
+                        commands.append([name, 'automember_remove_condition',
+                                         condition_args])
+
+                if exclusive is not None:
```
This test is not needed as you will be using `exclusive_*`.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,385 @@
+            elif state == "rebuild":
+                if res_find is None:
+                    ansible_module.fail_json(msg="No service '%s'" % name)
+                commands.append([None, 'automember_rebuild',
```
For `state: rebuilt` the module should accept a list of user and host names.

jake2184 replied:
This is the only comment I don't think I've fixed.

To accept a list of objects, what are you proposing? Adding a new module argument just for this?

rjeffman replied:
The CLI command automember-rebuild allows to rebuild automember rules for a list of users and/or hosts.

You should add twor parameters to the module `users` and `hosts` that are only valid when `state: rebuilt`. These parameters should be an optional list.

rjeffman commented on tests/automember/test_automember.yml:
```diff
@@ -0,0 +1,195 @@
+          expression: 'cn'
+    register: result
+    failed_when: result.changed or result.failed
+
```
Tests for `automember_type: hostgroup` with `action: member` are missing.

rjeffman commented on tests/automember/test_automember.yml:
```diff
@@ -0,0 +1,195 @@
+          expression: 'uid'
+    register: result
+    failed_when: result.changed or result.failed
+
```
Tests for `automember_type: group` with `action: member` are missing.

rjeffman reviewed (changes requested):
@jake2184 thank you for the work so far.

There are still a few things to be fixed so that we can merge the automember module.

Check the code review comments, and, please, add example playbooks (`playbooks/automember`) with rules, as the ones currently available do not have any. Also add examples with `action: member` in the README, and in the playbooks directory.

jake2184 commented:
I've responded to most comments, awaiting CI outcome of the tests

rjeffman reviewed (changes requested):
Now the only thing missing is the parameters for `state: rebuilt`.

Nicely done!

rjeffman commented on tests/automember/test_automember.yml:
```diff
@@ -0,0 +1,312 @@
+    register: result
+    failed_when: result.changed or result.failed
+
+  - name: Remove testgroup group automember rule member condition
```
This test is failing. Shouldn't it have `state: absent` ?

jake2184 commented:
Updated the tests, waiting for CI.

Also added "hosts" and "users" params to the rebuild module.

I'm not sure how multiple args should be passed to api.Command - I've just passed a list, will that work? 
``` 
                   commands.append([None, 'automember_rebuild',
                                    {"users": [
                                        to_text(_u)
                                        for _u in rebuild_users]}])

```

t-woerner commented:
Yes, it should be like this.

t-woerner commented:
test_automember.yml fails at the moment with 

TASK [Remove testgroup group automember rule member condition] *****************
task path: ansible-freeipa/tests/automember/test_automember.yml:138
fatal: [ipaserver.el83.local]: FAILED! => {"changed": false, "failed_when_result": true, "msg": "'dict' object has no attribute 'split'"}

t-woerner commented:
`inclusive: [{'key': 'manager', 'expression': 'mscott'}]` is a list of dicts.

jake2184 commented:
Thanks for your pointers - though it is on me to fix the tests. I don't have a nice way to run them locally so am using the CI to run the tests, so don't necessarily expect them all to pass first time! I'll update here when I have them clean and in need of review.

t-woerner commented:
I am using VMs to run the tests locally.

t-woerner commented:
What do you think about creating a standalone PR for https://github.com/freeipa/ansible-freeipa/pull/486/commits/9ba44a3b1d9b23c9a9c10b47ec3261bcdb779661 ?
We could use it also in another change before this PR is ready to be merged.

jake2184 commented:
Well personally I'd much rather wait and have this merged. I've been waiting for months for this PR, we're proactively working on it, and it's very nearly there. Basically once tests are passing I don't think there is anything outstanding and it could be merged?

t-woerner commented:
Ok, we will check the final PR again.

t-woerner commented:
Please add 2021 to the date and add yourself to authors in ipaautomember.py

t-woerner commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,397 @@
+# -*- coding: utf-8 -*-
+
+# Authors:
+#   Mark Hahl <mhahl@redhat.com>
```
Please add yourself to Authors

t-woerner commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,397 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import ipalib.errors
```
Please import `ipalib_errors` from `ansible_freeipa_module` and use `ipalib_errors` as a replacement of `ipalib.errors`.

t-woerner commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,397 @@
+
+    try:
+        _result = api_command(module, "automember_show", to_text(name), _args)
+    except ipalib.errors.NotFound:
```
Please use `ipalib_errors` from `ansible_freeipa_module` instead.

t-woerner commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,397 @@
+module: ipaautomember
+short description: Add and delete FreeIPA Auto Membership Rules.
+description:
+- Add, modify and delete an IPA Auto Membership Rules.
```
Please put this in the same line as `description:` or use 
```
description: |
  Add, modify and delete an IPA Auto Membership Rules.
```

jake2184 commented:
Any idea why this is failing on Centos7 and not Centos8? Looks like the same python version, is the FreeIPA version the same? Can someone trigger a re-run of the failed test (I get auth denied)

rjeffman commented on plugins/module_utils/ansible_freeipa_module.py:
```diff
@@ -307,7 +307,12 @@ def compare_args_ipa(module, args, ipa):  # noqa
+        if ignore is None:
+            ignore = []
+
+        filtered_args = [k for k in args.keys() if k not in ignore]
```
It is better not to use ".keys()" here, please, change it to `[key for key in args if key not in ignore]`.

The main reason is that we are trying to enable pylint verification, and it wil fail if kept the way it is now.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,404 @@
+                    exclusive_del = []
+
+                for _inclusive in inclusive_add:
+                    key = _inclusive[:_inclusive.find("=")]
+                    regex = _inclusive[_inclusive.find("=")+1:]
```
I suggest simplifying these two lines into one: `key, regex = _inclusive.split("=", 1)`.

I also suggest to abort if key or regex is empty.

jake2184 replied:
huh. Never knew about that very handy parameter.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,404 @@
+                                     condition_args])
+
+                for _inclusive in inclusive_del:
+                    key = _inclusive[:_inclusive.find("=")]
+                    regex = _inclusive[_inclusive.find("=")+1:]
```
same as above.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,404 @@
+                                     condition_args])
+
+                for _exclusive in exclusive_add:
+                    key = _exclusive[:_exclusive.find("=")]
+                    regex = _exclusive[_exclusive.find("=")+1:]
```
same as above.

rjeffman commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,404 @@
+                                     condition_args])
+
+                for _exclusive in exclusive_del:
+                    key = _exclusive[:_exclusive.find("=")]
+                    regex = _exclusive[_exclusive.find("=")+1:]
```
same as above.

rjeffman commented:
Tested on the CLI (not the module) and it should work with Centos 7. Just restarted the test.

All FreeIPA version are different in the tested environments. CentOS 7 has FreeIPA version 4.6.8, CentOS 8 is 4.8.7 and Fedora is 4.9.3.

rjeffman commented:
After re-running the tests, the same test still fails. Fixing it will require some debuging.

rjeffman commented:
I could reproduce your issue locally, with CentOS-7, but have no time to debug it.

If you have `podman` installed, use [this script](https://gist.github.com/rjeffman/58d3989519e19af3f618b20e11f6057f) to test in a container and debug (usually, I debug with `ansible_module.warn("VAR: %s" % var)`).

rjeffman commented:
For some reason, it looks like `key` and `automemberinclusiveregex` are swaped. I didn't look any further.

```
TASK [Add testhostgroup hostgroup automember rule member condition] ************************************
[WARNING]: testhostgroup automember_add_condition: {'automemberinclusiveregex': u'fqdn', 'type':
u'hostgroup', 'key': u'.*.domain.com'}
fatal: [dev-test-master]: FAILED! => {"changed": false, "failed_when_result": true, "msg": "automember_add_condition: testhostgroup: .*.domain.com is not a valid attribute."}
```

chr15p commented on plugins/modules/ipaautomember.py:
```diff
@@ -0,0 +1,404 @@
+    """Transform a list of dicts into a list with the format of key=value."""
+    transformed = []
+    for condition in conditions:
+        transformed.append('='.join(str(x) for x in condition.values()))
```
This assumes that python lays out dicts in memory in a consistent fashion. It doesn't.
It assumes that the first item in the condition dict is key, and the second item is expression so condition.values gives [key, expression] and this outputs key=expression,  this isn't true on my RHEL7 box (and I assume Centos7)   
For me conditions is 
` [{'expression': '.*.domain.com', 'key': 'fqdn'}]`
so we end up with transformed returning  ".*.domain.com'=fqdn"  then later we assume that its key=value  and so the key and expression get swapped causing the testing errors you are seeing 

as a rough guess rewriting this as 
`transformed.append('%s=%s'%(condition['key'], condition['expression']) )`
would fix the issue   (no promises it wouldn't cause others elsewhere!)

It looks like the implementation details of different versions of python are breaking stuff for you.

rjeffman commented:
@jake2184, sorry, but you'll need to rebase this PR.

For `plugins/module_utils/ansible_freeipa_module.py` just use what's on the `master` branch.

jake2184 commented:
Rebased, comments resolved, awaiting CI checks

rjeffman reviewed (approved):
LGTM.

rjeffman commented:
@jake2184, code for automember is good.

Please, merge all the commits into a single one, and use a commit message base on the one found on PR #357.

t-woerner commented on README.md:
```diff
@@ -12,6 +12,7 @@ Features
 * One-time-password (OTP) support for client installation
 * Repair mode for clients
 * Backup and restore, also to and from controller
+* Modules for automembership rule management
```
Please also add the reference to README-automember.md in the end of the file.

t-woerner commented on tests/automember/test_automember.yml:
```diff
@@ -0,0 +1,313 @@
+    register: result
+    failed_when: result.changed or result.failed
+
+######
```
Is this a comment?

t-woerner commented on README-automember.md:
````diff
@@ -0,0 +1,121 @@
+Example playbook to make sure group automember rule is present with no conditions.
+
+```yaml
+---
````
The examples are not consistent. Please always add 
```
---
- name: Playbook to ...
  hosts: ipaserver
  become: yes
```

t-woerner reviewed (approved):
LGTM

Wolskie commented:
@jake2184 thank you for taking over the work on this module, I appreciate it very much.
